kafka-plugins-0.10/src/main/java/io/cdap/plugin/batch/source/KafkaBatchSource.java

```diff
@@ -134,16 +136,20 @@ public KafkaBatchSource(Kafka10BatchConfig config) {
 
   @Override
   public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
-    config.validate();
-    pipelineConfigurer.getStageConfigurer().setOutputSchema(config.getSchema());
+    StageConfigurer stageConfigurer = pipelineConfigurer.getStageConfigurer();
+    FailureCollector failureCollector = stageConfigurer.getFailureCollector();
+    config.validate(failureCollector);
+    stageConfigurer.setOutputSchema(config.getSchema(failureCollector));
+    failureCollector.getOrThrowException();
   }
 
   @Override
   public void prepareRun(BatchSourceContext context) throws Exception {
     Job job = JobUtils.createInstance();
     Configuration conf = job.getConfiguration();
 
-    KafkaPartitionOffsets partitionOffsets = config.getInitialPartitionOffsets();
+    FailureCollector failureCollector = context.getFailureCollector();
+    KafkaPartitionOffsets partitionOffsets = config.getInitialPartitionOffsets(failureCollector);
```

Comment:
as mentioned before, do all the validation here before anything else happens. This should be:

```
    FailureCollector failureCollector = context.getFailureCollector();
    KafkaPartitionOffsets partitionOffsets = config.getInitialPartitionOffsets(failureCollector);
    Schema schema = config.getSchema(failureCollector);
    ... partitions = config.getPartitions(failureCollector);
    failureCollector.getOrThrowException();
```

Reply:
Ok, but I previously moved FailureCollector later in the function, since I didn't think the if block or the saveKafkaRequests call needed to be done after getOrThrowException.